### PR TITLE
Delay KPO pod clean up for Airflow to handle pod cleanup

### PR DESCRIPTION
**Pull request by abhishekbhakat**
Related Issue: #16

### CODE REVIEW

File: astronomer/cleanup_pods/command_line.py [hunk 1]
```diff
@@ -48,11 +51,23 @@ def cleanup(namespace):
...
            (pod_phase == POD_FAILED and pod_restart_policy == POD_RESTART_POLICY_NEVER) or
            (pod_reason == POD_REASON_EVICTED)):
-
+            if pod.metadata.labels.get('airflow_kpo_in_cluster'):
+                terminal_state_duration_mins = (
+                    (
+                        datetime.now(timezone.utc)
+                        - max(
+                            x.last_transition_time
+                            for x in pod.status.conditions
```
Comment (jedcunningham) [inline astronomer/cleanup_pods/command_line.py:65]: I wonder if we'd be better off iterating over the `pod.status.containerStatuses[].finishedAt`?

File: astronomer/cleanup_pods/command_line.py [hunk 2]
```diff
@@ -48,11 +51,23 @@ def cleanup(namespace):
...
         pod_reason = pod.status.reason.lower() if pod.status.reason else ''
         pod_restart_policy = pod.spec.restart_policy.lower()
+          
 
         if (pod_phase == POD_SUCCEEDED or
            (pod_phase == POD_FAILED and pod_restart_policy == POD_RESTART_POLICY_NEVER) or
            (pod_reason == POD_REASON_EVICTED)):
-
+            if pod.metadata.labels.get('airflow_kpo_in_cluster'):
+                terminal_state_duration_mins = (
```
Comment (jedcunningham) [inline astronomer/cleanup_pods/command_line.py]:
```suggestion
                terminal_state_duration = (
```

File: astronomer/cleanup_pods/command_line.py [hunk 3]
```diff
@@ -48,11 +51,23 @@ def cleanup(namespace):
...
+                    (
+                        datetime.now(timezone.utc)
+                        - max(
+                            x.last_transition_time
+                            for x in pod.status.conditions
+                        ).replace(tzinfo=None)
+                    )
+                ).total_seconds()
+                if terminal_state_duration_mins<=KPO_POD_DELETION_GRACE_PERIOD:
```
Comment (jedcunningham) [inline astronomer/cleanup_pods/command_line.py]:
```suggestion
                if terminal_state_duration <= KPO_POD_DELETION_GRACE_PERIOD:
```

File: astronomer/cleanup_pods/command_line.py (hunk 1 above)
Comment (abhishekbhakat) [reply astronomer/cleanup_pods/command_line.py:65]: I believe the property you are trying to point will only occur if the pod is in Terminating ?
Also, can you point me to the exact property ? I can't find it in [pod.status.container_status](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1ContainerStatus.md).

Comment (abhishekbhakat) [reply astronomer/cleanup_pods/command_line.py:65]: Isn't [last_transition_time](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1PodCondition.md#:~:text=last_transition_time) looks more safer?  As this tells when was the last status changed in pod.

Comment (jedcunningham) [reply astronomer/cleanup_pods/command_line.py:65]: Yes, sorry, it's actually `pod.status.containerStatuses[].state.terminated.finshedAt`.

Probably splitting hairs.

File: astronomer/cleanup_pods/command_line.py [hunk 4]
```diff
@@ -1,9 +1,12 @@
 import argparse
 import logging
+import os
 
 from kubernetes import client, config
 from kubernetes.client.rest import ApiException
 
+from datetime import datetime, timezone
+
```
Comment (jedcunningham) [inline astronomer/cleanup_pods/command_line.py]:
```suggestion
from datetime import datetime, timezone

from kubernetes import client, config
from kubernetes.client.rest import ApiException

```

ruff wants to reorder these imports

**Review by jedcunningham — COMMENTED** [review]
Overall the code change looks good. However, this breaks the [tests](https://github.com/astronomer/astronomer-airflow-scripts/blob/main/tests/test_cleanup_pods.py). Can you fix those and write a test for this change as well?

**Comment by abhishekbhakat** [comment]
> Overall the code change looks good. However, this breaks the [tests](https://github.com/astronomer/astronomer-airflow-scripts/blob/main/tests/test_cleanup_pods.py). Can you fix those and write a test for this change as well?

I made pre-existing tests not specific to KPO. And added a separate test for testing delay in deleting KPO pods.

File: astronomer/cleanup_pods/command_line.py (hunk 1 above)
Comment (abhishekbhakat) [reply astronomer/cleanup_pods/command_line.py:65]: Oh I see! But I have seen situations where finshedAt is null:

```yaml
initContainerStatuses:
  - state:
      terminated:
        exitCode: 103
        reason: Error
        startedAt: '2022-03-16T07:00:00Z'
        finishedAt: '2022-03-16T07:00:00Z'
        # ...
    lastState: {}
    ready: false
    restartCount: 0
    # ...
containerStatuses:
  - state:
      terminated:
        exitCode: 137
        reason: ContainerStatusUnknown
        message: The container could not be located when the pod was terminated
        startedAt: null
        finishedAt: null
    lastState: {}
    ready: false
    restartCount: 0
    started: false
# ...
```

Thoughts ?

File: tests/test_cleanup_pods.py [hunk 5]
```diff
@@ -80,8 +85,27 @@ def test_cleanup_api_exception_continue(load_incluster_config, list_namespaced_p
...
+@mock.patch('kubernetes.client.CoreV1Api.list_namespaced_pod')
+@mock.patch('kubernetes.config.load_incluster_config')
+def test_cleanup_kpo_pods_delay(load_incluster_config, list_namespaced_pod, delete_pod):
+    pod1 = MagicMock()
+    pod1.metadata.name = 'dummy'
+    pod1.metadata.labels = {'airflow_kpo_in_cluster': True}
+    pod1.status.phase = 'Succeeded'
+    pod1.status.reason = None
+    container1 = MagicMock()
+    container1.last_transition_time = datetime.now(timezone.utc) - timedelta(seconds=3599)
```
Comment (jedcunningham) [inline tests/test_cleanup_pods.py]: Let's leave a little more buffer here.
```suggestion
    container1.last_transition_time = datetime.now(timezone.utc) - timedelta(seconds=300)
```

File: tests/test_cleanup_pods.py [hunk 6]
```diff
@@ -80,8 +85,27 @@ def test_cleanup_api_exception_continue(load_incluster_config, list_namespaced_p
...
+    pod1.status.conditions = [container1]
+    list_namespaced_pod().items = [pod1]
+    cleanup('awesome-namespace')
+    delete_pod.assert_not_called()
+    load_incluster_config.assert_called_once()
```
Comment (jedcunningham) [inline tests/test_cleanup_pods.py:111]: Can we also check that after an hour (or whatever), we do end up deleting it?

Comment (abhishekbhakat) [reply tests/test_cleanup_pods.py:111]: Added another test for this.

**Review by jedcunningham — APPROVED** [review]

**Comment by jedcunningham** [comment]
Ran ruff and pytest locally.

**Pull request merged**